### DISCUSSION TITLE
Android RuleSet

### DISCUSSION THREAD

**Discussion by BraisGabin** [detekt/detekt#3303] [Ideas]
Lately we have some active contributors in the project that have an Android background. And there were some comments about some rules that for Android doesn't fit very well. Or some rules that only have sense in the Android context, etc. There were some comments about creating an Android rule set too but there were offtopic ideas in other PR/issues.

So, I will like to talk about the creation of a new RuleSet (or even multiple) for Android. I see multiple ways to do this:
- Add it as a default ruleset
- Add it as a first-party plugin
- Create it as a third-party plugin

I think that the first-party plugin have more sense. A plugin under the "detekt umbrella" but not in the core repository. My reasons:

- The users already know detekt so they will trust those rules more than in a third-party.
- In the plugin we can be more aggresive changing our rules. We know that android changes *a lot* so we don't need to be tied to the detekt release schedule.

Ideas for rules in the ruleset:
- Don't use the `Serializable` interface in a Bundle nor an Intent.
- Use `Intent(context, SomeActivity::class)` outside the companion object of `SomeActivity`. (Use the pattern `SomeActivity.getCallingIntent()`)
- Call `SomeFragment()` constructor outside the companion object of `SomeFragment`. (Use the pattern `SomeFragment.newInstance()`)
- Use AndroidX Compat instead of the framework. We will need to maintain a big list here.
- Activity that extends directly from `Activity` instead of `AppCompatActivity`
- ...

**Comment by picklebento** [comment]
Adding a first-party plugin makes sense. Note that we should be able to add the dependency on Android ruleset when we are applying the android flavor of `detekt-gradle-plugin`.

The idea pool is even bigger: [Android lint checks](https://cs.android.com/android-studio/platform/tools/base/+/mirror-goog-studio-master-dev:lint/libs/lint-checks/src/main/java/com/android/tools/lint/checks/). Since Android Lint has to support all Android file types, including Java, Kotlin, and XML, running a new rule in Android lint is generally slower than a detekt rule. In the repo I am working on, we run detekt as a git-commit hook (Usually 2s for the changed files), but we run Android Lint only on CI (Because it takes 10 minutes to run).

**Reply by BraisGabin** [reply]
I missed this! Sure, that's a great pool of ideas.

**Comment by cortinico** [comment]
I think that's a great idea.

I also recall having a lot of discussions for Android-specific configuration of plain detekt rules (something related to https://github.com/detekt/detekt/issues/2231) Are we planning to address this as well?

**Reply by BraisGabin** [reply]
It sounds good! I don't know if detekt have apis for that but we could add them.

**Comment by BraisGabin** [comment]
One rule that I miss:

```kotlin
Intent(context, XXXActivity::class.java)
```

This code should be only found inside the `XXXActivity` class. Usually in a companion object. But no one outside an Activity should know how to call it. XXXActivity should provide a IntentFactory or something similar to get the needed `Intent` to call it.
